src/nvim/message.c

```diff
@@ -1012,18 +1020,56 @@ static void msg_hist_add(const char *s, int len, int hl_id)
     text.size--;
   }
   if (text.size == 0) {
-    return;
+    return -1;
   }
   text.data = xmemdupz(s, text.size);
 
   HlMessage msg = KV_INITIAL_VALUE;
   kv_push(msg, ((HlMessageChunk){ text, hl_id }));
-  msg_hist_add_multihl(msg, false);
+  MsgID id = msg_hist_add_multihl(0, msg, false, NULL);
+  return id;
 }
 
 static bool do_clear_hist_temp = true;
 
-static void msg_hist_add_multihl(HlMessage msg, bool temp)
+/// returns message history item based on it's id or NULL if not found
+static MessageHistoryEntry *msg_find_by_id(MsgID id)
+{
+  if (id <= 0) {
+    return NULL;
+  }
+  MessageHistoryEntry *entry = msg_hist_last;
+  while (entry != NULL && entry->msg_id != id) {
+    entry = entry->prev;
+  }
+  return entry;
+}
+
+static void emit_progress_event(MessageHistoryEntry *msg)
```

Comment:
this is the C codebase convention, AFAICT. (the `_event` suffix is for the event-loop, something totally different)
```suggestion
static void do_autocmd_progress(MessageHistoryEntry *msg)
```